File: desktop/packages/core/test/data/subject-repository.test.ts
import { afterEach, beforeEach, expect, test } from 'vitest';

import { SqliteSubjectRepository } from '../../src/data/subject-repository.js';
import type { Database } from '../../src/db/connection.js';
import { ensureLocalUser } from '../../src/db/local-user.js';
import { localUserId } from '../../src/ids.js';
import { toEpochSeconds } from '../../src/time.js';
import { fixedClock, freshDb } from '../support/fixture.js';

const t0 = new Date('2026-07-26T09:00:00.000Z');

let db: Database;
let repo: SqliteSubjectRepository;
let clock: ReturnType<typeof fixedClock>;

beforeEach(() => {
  db = freshDb();
  ensureLocalUser(db);
  clock = fixedClock(t0);
  repo = new SqliteSubjectRepository(db, clock.now);
});
afterEach(() => {
  db.close();
});

test('create inserts a row visible to list', async () => {
  await repo.create({ name: 'Physics' });
  const subjects = await repo.list();
  expect(subjects).toHaveLength(1);
  expect(subjects[0]?.name).toBe('Physics');
});

test('create attaches the row to the local user', async () => {
  // Locked decision 4: every user_id FK is populated from day one so adding a
  // server later needs no migration.
  await repo.create({ name: 'Physics' });
  expect(db.prepare('SELECT user_id FROM subjects').get<{ user_id: string }>()?.user_id).toBe(
    localUserId,
  );
});

test('list excludes soft-deleted rows', async () => {
  const id = await repo.create({ name: 'Old' });
  db.prepare('UPDATE subjects SET deleted_at = ? WHERE id = ?').run(toEpochSeconds(t0), id);
  expect(await repo.list()).toEqual([]);
});

test('create persists colour, source and source name; defaults are self/null', async () => {
  const plainId = await repo.create({ name: 'Maths' });
  const richId = await repo.create({
    name: 'Flutter',
    color: '#42A5F5',
    source: 'course',
    sourceName: 'Udemy',
  });
  const rows = await repo.list();
  const plain = rows.find((s) => s.id === plainId);
  const rich = rows.find((s) => s.id === richId);
  // Asymmetric on purpose: three identical values could not catch two swapped
  // columns.
  expect([plain?.color, plain?.source, plain?.sourceName]).toEqual([null, 'self', null]);
  expect([rich?.color, rich?.source, rich?.sourceName]).toEqual(['#42A5F5', 'course', 'Udemy']);
});

test('update rewrites all editable fields, and null clears them', async () => {
  const id = await repo.create({
    name: 'Chem',
    color: '#EF5350',
    source: 'school',
    sourceName: 'MRSM',
  });
  await repo.update(id, { name: 'Chemistry', color: null, source: 'self', sourceName: null });
  const s = (await repo.list())[0];
  expect([s?.name, s?.color, s?.source, s?.sourceName]).toEqual([
    'Chemistry',
    null,
    'self',
    null,
  ]);
});

test('update touches only the named row', async () => {
  const a = await repo.create({ name: 'A' });
  await repo.create({ name: 'B' });
  await repo.update(a, { name: 'A2', color: null, source: 'self', sourceName: null });
  const names = (await repo.list()).map((s) => s.name).sort();
  expect(names).toEqual(['A2', 'B']);
});

test('archiving moves a subject between the two lists, both ways', async () => {
  const id = await repo.create({ name: 'Physics' });
  await repo.setArchived(id, true);
  expect(await repo.list()).toEqual([]);
  expect((await repo.list({ archived: true }))[0]?.name).toBe('Physics');
  await repo.setArchived(id, false);
  expect((await repo.list())[0]?.name).toBe('Physics');
  expect(await repo.list({ archived: true })).toEqual([]);
});

test('archived is exposed as a boolean, not the stored 0/1', async () => {
  const id = await repo.create({ name: 'Physics' });
  expect((await repo.list())[0]?.archived).toBe(false);
  await repo.setArchived(id, true);
  expect((await repo.list({ archived: true }))[0]?.archived).toBe(true);
});

test('remove is soft: gone from both lists, the row survives', async () => {
  const id = await repo.create({ name: 'Physics' });
  await repo.remove(id);
  expect(await repo.list()).toEqual([]);
  expect(await repo.list({ archived: true })).toEqual([]);
  const row = db.prepare('SELECT deleted_at FROM subjects').get<{ deleted_at: number | null }>();
  expect(row?.deleted_at, 'soft delete, never DELETE — history joins keep the name').not.toBeNull();
});

test('a soft-deleted archived subject stays hidden from both lists', async () => {
  const id = await repo.create({ name: 'Physics' });
  await repo.setArchived(id, true);
  await repo.remove(id);
  expect(await repo.list()).toEqual([]);
  expect(await repo.list({ archived: true })).toEqual([]);
});

test('every write advances the updated_at watermark', async () => {
  // updated_at is the liveness watermark crash recovery bounds a session by, so
  // a write path that forgets it is a real defect, not cosmetic.
  const id = await repo.create({ name: 'Physics' });
  const readUpdatedAt = (): number =>
    db.prepare('SELECT updated_at u FROM subjects WHERE id = ?').get<{ u: number }>(id)?.u ?? 0;
  const created = readUpdatedAt();
  expect(created).toBe(toEpochSeconds(t0));

  for (const [minutes, write] of [
    [10, () => repo.update(id, { name: 'P', color: null, source: 'self', sourceName: null })],
    [20, () => repo.setArchived(id, true)],
    [30, () => repo.remove(id)],
  ] as const) {
    const then = new Date(t0.getTime() + minutes * 60_000);
    clock.set(then);
    await write();
    expect(readUpdatedAt(), `after the write at +${String(minutes)}min`).toBe(
      toEpochSeconds(then),
    );
  }
});

test('list is newest-first', async () => {
  const a = await repo.create({ name: 'A' });
  clock.set(new Date(t0.getTime() + 60 * 60_000));
  const b = await repo.create({ name: 'B' });
  expect((await repo.list()).map((s) => s.id)).toEqual([b, a]);
});

test('list breaks same-second ties deterministically', async () => {
  // created_at has second precision, so without the id tiebreak two subjects
  // made in the same second would order arbitrarily between runs. newId() is
  // UUIDv7, so descending id is descending time.
  const first = await repo.create({ name: 'A' });
  const second = await repo.create({ name: 'B' });
  expect((await repo.list()).map((s) => s.id)).toEqual([second, first]);
});

test('createdAt comes back as a Date at second precision', async () => {
  await repo.create({ name: 'Physics' });
  const s = (await repo.list())[0];
  expect(s?.createdAt).toBeInstanceOf(Date);
  expect(s?.createdAt.getTime()).toBe(t0.getTime());
});

File: app-flutter/test/features/subjects/subject_repository_test.dart
import 'package:drift/drift.dart' show Value;
import 'package:drift/native.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:nerdyapp/core/db/database.dart';
import 'package:nerdyapp/core/db/local_user.dart';
import 'package:nerdyapp/features/session/data/session_repository.dart';
import 'package:nerdyapp/features/session/domain/active_session.dart';
import 'package:nerdyapp/features/subjects/data/subject_repository.dart';

void main() {
  late AppDatabase db;
  late SubjectRepository repo;

  setUp(() async {
    db = AppDatabase(NativeDatabase.memory());
    await ensureLocalUser(db);
    repo = SubjectRepository(db);
  });
  tearDown(() => db.close());

  test('createSubject inserts a row visible to watchSubjects', () async {
    await repo.createSubject('Physics');
    final subjects = await repo.watchSubjects().first;
    expect(subjects, hasLength(1));
    expect(subjects.single.name, 'Physics');
  });

  test('watchSubjects excludes soft-deleted rows', () async {
    final id = await repo.createSubject('Old');
    await (db.update(db.subjects)..where((s) => s.id.equals(id))).write(
        SubjectsCompanion(deletedAt: Value(DateTime.now().toUtc())));
    final subjects = await repo.watchSubjects().first;
    expect(subjects, isEmpty);
  });

  test('create persists colour, source and source name; defaults are self/null',
      () async {
    final plainId = await repo.createSubject('Maths');
    final richId = await repo.createSubject('Flutter',
        color: '#42A5F5', source: 'course', sourceName: 'Udemy');
    final rows = await repo.watchSubjects().first;
    final plain = rows.singleWhere((s) => s.id == plainId);
    final rich = rows.singleWhere((s) => s.id == richId);
    expect((plain.color, plain.source, plain.sourceName), (null, 'self', null));
    expect((rich.color, rich.source, rich.sourceName),
        ('#42A5F5', 'course', 'Udemy'));
  });

  test('update rewrites all editable fields', () async {
    final id = await repo.createSubject('Chem',
        color: '#EF5350', source: 'school', sourceName: 'MRSM');
    await repo.updateSubject(id,
        name: 'Chemistry', color: null, source: 'self', sourceName: null);
    final s = (await repo.watchSubjects().first).single;
    expect((s.name, s.color, s.source, s.sourceName),
        ('Chemistry', null, 'self', null));
  });

  test('archiving removes from the active list but not from history joins',
      () async {
    final id = await repo.createSubject('Physics');
    // An ended session, so the history join has something to find.
    final session = ActiveSession.start(
        id: 'sess-arch', subjectId: id, startedAt: DateTime.utc(2026, 7, 26));
    final sessions = SessionRepository(db);
    await sessions.insertStartedSession(session);
    await sessions.endSession(
      id: session.id,
      endedAt: session.startedAt.add(const Duration(minutes: 30)),
      actualDuration: const Duration(minutes: 30),
      totalPaused: Duration.zero,
    );

    await repo.setArchived(id, archived: true);
    expect(await repo.watchSubjects().first, isEmpty);
    expect((await repo.watchSubjects(archived: true).first).single.name,
        'Physics');
    expect(
        (await sessions.watchHistory().first).single.subjectName, 'Physics');

    await repo.setArchived(id, archived: false);
    expect((await repo.watchSubjects().first).single.name, 'Physics');
  });

  test('delete is soft: gone from both lists, history keeps the name',
      () async {
    final id = await repo.createSubject('Physics');
    final session = ActiveSession.start(
        id: 'sess-del', subjectId: id, startedAt: DateTime.utc(2026, 7, 26));
    final sessions = SessionRepository(db);
    await sessions.insertStartedSession(session);
    await sessions.endSession(
      id: session.id,
      endedAt: session.startedAt.add(const Duration(minutes: 30)),
      actualDuration: const Duration(minutes: 30),
      totalPaused: Duration.zero,
    );

    await repo.deleteSubject(id);
    expect(await repo.watchSubjects().first, isEmpty);
    expect(await repo.watchSubjects(archived: true).first, isEmpty);
    final row = await db.select(db.subjects).getSingle();
    expect(row.deletedAt, isNotNull); // soft, not DELETE
    expect(
        (await sessions.watchHistory().first).single.subjectName, 'Physics');
  });
}
